src/org/purefn/river/batch.clj

```diff
@@ -0,0 +1,137 @@
+(ns org.purefn.river.batch
+  "A consumer suitable for batch reading operations from Kafka."
+  (:require [clojure.pprint :as pprint]
+            [clojure.spec.alpha :as s]
+            [com.stuartsierra.component :as component]
+            ;; temp, remove
+            [org.purefn.river.serdes.nippy :as serdes]
+            ;;----
+            [taoensso.timbre :as log])
+  (:import [org.apache.kafka.clients.consumer KafkaConsumer]
+           [org.apache.kafka.common TopicPartition]
+           [org.apache.kafka.common.errors WakeupException]
+           [org.purefn.river Message]))
+
+(defn kafka-consumer
+  [group-id servers]
+  (KafkaConsumer. {"auto.offset.reset" "earliest"
+                   "bootstrap.servers" servers
+                   "enable.auto.commit" false
+                   "group.id" group-id
+                   "client.id" group-id}
+                  (serdes/nippy-deserializer)
+                  (serdes/nippy-deserializer)))
+
+(defn create-consumers
+  [{:keys [::topics ::group-id ::threads ::bootstrap-servers] :as config}]
+  (log/info config)
+  (->> (range threads)
+       (map (fn [_]
+              (log/info "Creating consumer" {:group-id group-id
+                                             :topics topics})
+                (doto (kafka-consumer group-id bootstrap-servers)
+                  (.subscribe topics))))))
+
+(defn process
+  [^KafkaConsumer consumer
+   ^clojure.lang.Atom closing
+   {:keys [::timeout] :as config}
+   dependencies
+   process-fn]
+  (let [commit #(.commitAsync consumer)
+        pfn (if (= 4 (->> (meta process-fn)
```

Comment:
Not sure what is going on here, should this scalar be defined...

Reply:
I'll document it, and probably instrument the call/spec the component constructor.  `process-fn` should be either 3-arity, or 4-arity.  4-arity if it needs to accept additional component dependencies (passed in via `component/using`).

At least that's my current thinking, open to better suggestions at this point...